ARM: dts: ux500: samsung-golden: Add magnetometer

AffeAli opened the pull request:
While writing a driver for the magnetometer I noticed that the ak8974 driver is compatible with it.
A mount-matix may be required.(I don't know how to get this information)

AffeAli committed:
dt-bindings: magnetometer: ak8974: Add Alps hscdtd008a

AffeAli committed:
iio: magnetometer: ak8974: add Alps hscdtd008a

The hscdtd008a is similar to the AK8974: only the whoami value and some registers are different.

AffeAli committed:
ARM: dts: ux500: samsung-golden: Add magnetometer

AffeAli committed:
dt-bindings: vendor-prefixes: Add Alps

stephan-gh commented on drivers/iio/magnetometer/ak8974.c:
```diff
@@ -491,6 +492,10 @@ static int ak8974_detect(struct ak8974 *ak8974)
 			 "detected %s, FW ver %02x, S/N: %04x\n",
 			 name, fw, sn);
 		break;
+	case AK8974_WHOAMI_VALUE_HSCDTD008A:
```
I would move this below `case AK8974_WHOAMI_VALUE_AK8974:`. Then the switch statement is ordered by WHOAMI.

stephan-gh commented on drivers/iio/magnetometer/ak8974.c:
```diff
@@ -685,6 +690,8 @@ static bool ak8974_writeable_reg(struct device *dev, unsigned int reg)
 		if (ak8974->variant == AK8974_WHOAMI_VALUE_AK8974)
 			return true;
+		if (ak8974->variant == AK8974_WHOAMI_VALUE_HSCDTD008A)
+			return true;
```
You could combine this with an OR above.

stephan-gh reviewed (changes requested):
Thanks again for the patches and especially for finding the similarity with ak8974!  
It seems to be working fine for me.

I compared the registers used by the mainline driver with the one listed in the Alps datasheet. They are almost identical, but there are some subtle differences that we should handle in the mainline driver. I tried to mark them inline:

```diff
diff --git a/drivers/iio/magnetometer/ak8974.c b/drivers/iio/magnetometer/ak8974.c
index e2b31e9d4b35..12fdf10eb916 100644
--- a/drivers/iio/magnetometer/ak8974.c
+++ b/drivers/iio/magnetometer/ak8974.c
@@ -54,15 +54,15 @@
 #define AK8974_DATA_X		0x10
 #define AK8974_DATA_Y		0x12
 #define AK8974_DATA_Z		0x14
-#define AK8974_INT_SRC		0x16
+#define AK8974_INT_SRC		0x16	// NOPE
 #define AK8974_STATUS		0x18
-#define AK8974_INT_CLEAR	0x1A
+#define AK8974_INT_CLEAR	0x1A	// NOPE
 #define AK8974_CTRL1		0x1B
 #define AK8974_CTRL2		0x1C
 #define AK8974_CTRL3		0x1D
-#define AK8974_INT_CTRL		0x1E
+#define AK8974_INT_CTRL		0x1E	// NOPE, CTRL4
 #define AK8974_INT_THRES	0x26  /* Absolute any axis value threshold */
-#define AK8974_PRESET		0x30
+#define AK8974_PRESET		0x30	// NOPE
 
 /* AK8974-specific offsets */
 #define AK8974_OFFSET_X		0x20
@@ -116,14 +116,14 @@
 
 #define AK8974_STATUS_DRDY	BIT(6) /* Data ready */
 #define AK8974_STATUS_OVERRUN	BIT(5) /* Data overrun */
-#define AK8974_STATUS_INT	BIT(4) /* Interrupt occurred */
+#define AK8974_STATUS_INT	BIT(4) /* Interrupt occurred */	// NOPE
 
 #define AK8974_CTRL1_POWER	BIT(7) /* 0 = standby; 1 = active */
-#define AK8974_CTRL1_RATE	BIT(4) /* 0 = 10 Hz; 1 = 20 Hz	 */
+#define AK8974_CTRL1_RATE	BIT(4) /* 0 = 10 Hz; 1 = 20 Hz	 */ // not quite
 #define AK8974_CTRL1_FORCE_EN	BIT(1) /* 0 = normal; 1 = force	 */
-#define AK8974_CTRL1_MODE2	BIT(0) /* 0 */
+#define AK8974_CTRL1_MODE2	BIT(0) /* 0 */ // NOPE
 
-#define AK8974_CTRL2_INT_EN	BIT(4)  /* 1 = enable interrupts	      */
+#define AK8974_CTRL2_INT_EN	BIT(4)  /* 1 = enable interrupts	      */ // NOPE
 #define AK8974_CTRL2_DRDY_EN	BIT(3)  /* 1 = enable data ready signal */
 #define AK8974_CTRL2_DRDY_POL	BIT(2)  /* 1 = data ready active high   */
 #define AK8974_CTRL2_RESDEF	(AK8974_CTRL2_DRDY_POL)
```

We can use the reported `ak8974->variant` to add special handling for some of the registers. I would:

1. Update `ak8974_writeable_reg` to disallow writing to registers that don't exist.

2. `AK8974_INT_SRC` and `AK8974_INT_CLEAR` are only used read-only. Apparently they don't cause any issues, so I guess we could leave the following two reads as-is: (not sure)

    https://github.com/novathor-mainline/linux/blob/5dadd3d12d57562a272b849777b1acaa2cc5d2c0/drivers/iio/magnetometer/ak8974.c#L352-L354

    https://github.com/novathor-mainline/linux/blob/5dadd3d12d57562a272b849777b1acaa2cc5d2c0/drivers/iio/magnetometer/ak8974.c#L284-L287

3. `AK8974_INT_CTRL` actually refers to a different register ("CTRL4"). We shouldn't write the interrupt configuration to it. Looking at [`hscdtd008a_i2c.c` the downstream driver writes 0x90](https://github.com/novathor-mainline/android_kernel/blob/9508895090960012f916561b2ede5eaf8124fabd/drivers/sensor/alps/hscdtd008a_i2c.c#L348-L350).

4. `AK8974_PRESET` is not needed at all, I guess.

5. `AK8974_CTRL1_RATE` has 2 bits instead (more possible values). Not sure if we need to handle it somewhere.

6. `AK8974_CTRL2_INT_EN` means something different (FIFO enable). The downstream driver does not write to CTRL2 at all, so I don't think we want to enable the FIFO.

These are just some thoughts, please let me know if you want to discuss something.

AffeAli commented:
I changed the simple stuff.
With the registers I tried tons of different variations (mostly with `AK8974_INT_CTRL` and `AK8974_PRESET`) of (not) writing  to them and nothing worked. At least for now I'm out of ideas on this.

Minecrell commented:
@AffeAli Does "nothing worked" mean reading values fails? Initialization fails? What is broken exactly?

AffeAli commented:
Sorry about my imprecise answer, I was a bit frustrated. Not writing to `AK8974_INT_CTRL` makes the reset fail. With the following patch (writing 0x90 to `AK8974_INT_CTRL` and not writing to `AK8974_PRESET`) the sensor values seem like they sometimes overflow. (They are near the 16-bit integer max value)
```
diff --git a/drivers/iio/magnetometer/ak8974.c b/drivers/iio/magnetometer/ak8974.c
index e2b31e9d4b35..dbbef8b8a3ef 100644
--- a/drivers/iio/magnetometer/ak8974.c
+++ b/drivers/iio/magnetometer/ak8974.c
@@ -242,10 +242,17 @@ static int ak8974_reset(struct ak8974 *ak8974)
        ret = regmap_write(ak8974->map, AK8974_CTRL3, AK8974_CTRL3_RESDEF);
        if (ret)
                return ret;
-       ret = regmap_write(ak8974->map, AK8974_INT_CTRL,
-                          AK8974_INT_CTRL_RESDEF);
-       if (ret)
-               return ret;
+       if (ak8974->variant != AK8974_WHOAMI_VALUE_HSCDTD008A) {
+               ret = regmap_write(ak8974->map, AK8974_INT_CTRL,
+                                  AK8974_INT_CTRL_RESDEF);
+               if (ret)
+                       return ret;
+       } else {
+               ret = regmap_write(ak8974->map, AK8974_INT_CTRL,
+                                  0x90);
+               if (ret)
+                       return ret;
+       }
 
        /* After reset, power off is default state */
        return ak8974_set_power(ak8974, AK8974_PWR_OFF);
@@ -268,6 +275,9 @@ static int ak8974_configure(struct ak8974 *ak8974)
                if (ret)
                        return ret;
        }
+       if (ak8974->variant == AK8974_WHOAMI_VALUE_HSCDTD008A) {
+               return 0;
+       }
        ret = regmap_write(ak8974->map, AK8974_INT_CTRL, AK8974_INT_CTRL_POL);
        if (ret)
                return ret;
@@ -679,20 +689,18 @@ static bool ak8974_writeable_reg(struct device *dev, unsigned int reg)
        case AK8974_INT_CTRL:
        case AK8974_INT_THRES:
        case AK8974_INT_THRES + 1:
+               return true;
        case AK8974_PRESET:
        case AK8974_PRESET + 1:
-               return true;
+               return ak8974->variant != AK8974_WHOAMI_VALUE_HSCDTD008A;
        case AK8974_OFFSET_X:
        case AK8974_OFFSET_X + 1:
        case AK8974_OFFSET_Y:
```

Minecrell commented:
> Sorry about my imprecise answer, I was a bit frustrated. 

I know that feeling :) It happens occasionally when working on kernel development, I don't think there is any way to avoid it. :/

> Not writing to AK8974_INT_CTRL makes the reset fail. With the following patch (writing 0x90 to AK8974_INT_CTRL and not writing to AK8974_PRESET) the sensor values seem like they sometimes overflow. (They are near the 16-bit integer max value)

Curious, does it really behave differently than before? I thought I saw negative values returned by  `in_magn_*_raw`, but I re-tested without your last changes and still get the 65000+ numbers.

AffeAli commented:
I just tested my changes again and I never got any negative values. 65000+ are common with the patch in my comment, but with the code in the PR only when holding a magnet close to the phone. Without a magnet it is always within ~5-500.

stephan-gh commented on drivers/iio/magnetometer/ak8974.c:
```diff
@@ -241,10 +242,17 @@ static int ak8974_reset(struct ak8974 *ak8974)
+			return ret;
+	} else {
+		ret = regmap_write(ak8974->map, AK8974_INT_CTRL,
+				   0x90);
```
To make this less confusing it might help if you create extra constants for these two:

e.g.

```diff
diff --git a/drivers/iio/magnetometer/ak8974.c b/drivers/iio/magnetometer/ak8974.c
index e2b31e9d4b35..9d07605ba623 100644
--- a/drivers/iio/magnetometer/ak8974.c
+++ b/drivers/iio/magnetometer/ak8974.c
@@ -141,6 +141,12 @@
 #define AK8974_INT_CTRL_PULSE	BIT(1) /* 0 = latched; 1 = pulse (50 usec) */
 #define AK8974_INT_CTRL_RESDEF	(AK8974_INT_CTRL_XYZEN | AK8974_INT_CTRL_POL)
 
+/* HSCDTD008A-specific control register */
+#define HSCDTD008A_CTRL4	0x1E
+#define HSCDTD008A_CTRL4_MMD	BIT(7)	/* must be set to 1 */
+#define HSCDTD008A_CTRL4_RANGE	BIT(4)	/* 0 = 14-bit output; 1 = 15-bit output */
+#define HSCDTD008A_CTRL4_RESDEF	(HSCDTD008A_CTRL4_MMD | HSCDTD008A_CTRL4_RANGE)
+
 /* The AMI305 has elaborate FW version and serial number registers */
 #define AMI305_VER		0xE8
 #define AMI305_SN		0xEA
```

then use `HSCDTD008A_CTRL4` and `HSCDTD008A_CTRL4_RESDEF` (= 0x90) in this function call.

stephan-gh commented on drivers/iio/magnetometer/ak8974.c:
```diff
@@ -267,6 +275,9 @@ static int ak8974_configure(struct ak8974 *ak8974)
 	}
+	if (ak8974->variant == AK8974_WHOAMI_VALUE_HSCDTD008A) {
+		return 0;
+	}
```
no need for brackets here

stephan-gh commented on arch/arm/boot/dts/ste-ux500-samsung-golden.dts:
```diff
@@ -50,6 +50,26 @@
+			reg = <0x0c>;
+
+			dvdd-supply = <&ab8500_ldo_aux1_reg>;
+			avdd-supply = <&ab8500_ldo_aux8_reg>;
```
Check the schematics to see which regulator belongs to which supply :)

AffeAli replied:
I was convinced I had them right this time :(

AffeAli commented:
Should I split your cast-to-s16 fix into a separate commit, as it is relevant for the entire driver?

Minecrell commented:
> Should I split your cast-to-s16 fix into a separate commit, as it is relevant for the entire driver?

I was going to send that upstream separately very soon(TM).

stephan-gh committed:
FROMLIST: iio: magnetometer: ak8974: Fix negative raw values in sysfs

At the moment, reading from in_magn_*_raw in sysfs tends to return
large values around 65000, even though the output of ak8974 is actually
limited to ±32768. This happens because the value is never converted
to the signed 16-bit integer variant.

Add an explicit cast to s16 to fix this.

Cc: Linus Walleij <linus.walleij@linaro.org>
Fixes: 7c94a8b2ee8c ("iio: magn: add a driver for AK8974")
Signed-off-by: Stephan Gerhold <stephan@gerhold.net>

Minecrell commented:
Tested on my samsung-golden. Seems to work fine.

I made a separate patch for the `(s16)` fix and sent it upstream. Should get applied in a couple of days. Then you can send your patches on top of it :)

Thanks again!
